improvement: permission explanations

ericli3690 opened the pull request:
Adds permission explanation screen to OS settings to explain to the user why we need the permissions we request, via a new AllPermissionsExplanationFragment fragment hosted in a new PermissionsExplanationActivity. Arthur requested that I create a new activity for this purpose.

## Purpose / Description
- Launches the above activity via an intent filter defined in AndroidManifest.xml, as per [the docs](https://developer.android.com/training/permissions/explaining-access#privacy-dashboard).
- Added comments to AndroidManifest.xml to flag that future maintainers should add any new runtime permissions to AllPermissionsExplanationFragment.
- Moved the logic for requesting external storage from the children of PermissionsFragment to the PermissionsFragment abstract class itself. This allows the methods to be called in both the PermissionsFragment children and the AllPermissionsExplanationFragment, minimizing code duplication.
- Added strings for telling the user that they need to grant or revoke permissions manually from the OS settings, strings for the permissions explanations, etc.
- Increased the amount of maximum lines in a PermissionItem's description so I could fit a longer explanation of why AnkiDroid needs notification permissions.
- Added POST_NOTIFICATIONS and RECORD_AUDIO to constants.xml so they can be referred to by AllPermissionsExplanation.

How the permissions explanation screen looks on a Play Store build:
<img width="767" height="820" alt="image" src="https://github.com/user-attachments/assets/887d85ce-b8ea-4462-8793-c8c6d279c52b" />

An example of the permissions explanation screen on a full debug build (not Play Store) (the strings in this screenshot are outdated, see the above screenshot for updated strings):
<img width="1121" height="413" alt="tablet_fulldebug_api30plus" src="https://github.com/user-attachments/assets/176b6d5c-54bc-486b-86b1-4cf9f1bcea95" />

The snackbar that shows up when you try to revoke a permission:
<img width="341" height="767" alt="api30plus_redirect_for_revoke" src="https://github.com/user-attachments/assets/4cdd5147-4b0c-4aa6-8da6-b12313425040" />

## Fixes
- Requested by Brayan. It's best to explain why we need permissions not only when we are requesting them in-app, but also whenever the user goes to check on them in the OS settings.

## Approach
- Creates a new activity as requested by Arthur.
- Based on my understanding of the app's permissions from reading `AndroidManifest`, it seems there are only really three kinds of permissions that we need to request from the user: externally managed storage, notifications, and microphone. Legacy storage permissions are only available until API 29, and this permissions explanation screen only shows up on API 31+. **Please suggest wording changes if you dislike any wording I've used.**

## How Has This Been Tested?
- Physical Samsung S23, API 34
- Emulated Pixel 3a, API 28
- Emulated Pixel 5, API 30
- Emulated Pixel 6a, API 33

## Checklist
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [x] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

<!--- Uncomment this section ONLY if this PR introduces new resources (external libraries, icons etc)
## Licenses
_For each new external resource, add a row to the table below:_

| Library | Description | License |
| --- | --- | --- |
| Sample Icon Library | Sample Description | [The Apache Software License, Version 2.0](http://www.apache.org/licenses/LICENSE-2.0.txt) |

**Maintainers:**

* [ ] Add the https://github.com/ankidroid/Anki-Android/labels/Licenses label
* [ ] Update the [licenses](https://github.com/ankidroid/Anki-Android/wiki/Licences) wiki when merging
--->

ericli3690 commented:
Need people to review. Please let me know if:

1. You want any UI changes. Right now, it's very basic.
2. You think any wording is unclear or incorrect.
3. I've missed any permissions that should be on this list.
4. You know how I can test the managed-storage permissions. Is there some setting in the emulator I can tick to make it think I've installed a GitHub / F-Droid build, for instance?

Haven't tested on different API versions yet. The above image is how it looks on my phone, which is a Samsung S23, API 34. I wanted to get UX / implementation feedback first before committing more time to testing.

ericli3690 commented:
Oops, forgot the copyright header and had to add the new activity to the ActivityList for tests. Should be fixed now.

BrayanDSO reviewed (commented):
Could you try reusing PermissionsActivity? Adding these permissions under an `Optional` section there should work.

ericli3690 commented:
Refactored to use PermissionsActivity. Refactored the permission fragments to minimize code duplication, fixed some issues. See the PR description for an updated changelog.

I've done more testing and things seem to be working. The only slightly weird bit of behaviour is that if the user clicks on a permission item to enable it, but then hits "do not grant" in the dialog, the fragment thinks the dialog to enable permissions failed and opens the OS settings. I can't think of a way to stop this from happening, but I think it's actually alright behaviour. If the user clicks on a permission item that's currently disabled, then we can assume that they want to grant that permission, and so it makes sense that we show them the ability to toggle the permission in the OS settings if they immediately click "do not grant" on the dialog.

Important question: Does anyone know how I can test the externally-managed-storage permissions? I haven't been able to test these yet.

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/permissions/AllPermissionsExplanation.kt:
```diff
@@ -0,0 +1,120 @@
+                offerToGrantOrRevokeOnClick(Permissions.postNotification)
+            }
+        }
+        // Microphone
+        // Minimum API 23, which is less than AnkiDroid's minimum targeted API version, so we show it unconditionally
+        recordAudioPermission.apply {
+            isVisible = true
+            offerToGrantOrRevokeOnClick(Permissions.recordAudioPermission)
+        }
```
We currently show this when a user wants to record audio. 

We /may/ want to inform users, but I feel it's just additional cognitive overhead here. 
We don't need mic access to run the app

david-allison commented on AnkiDroid/src/main/res/values/12-dont-translate.xml:
```diff
@@ -42,5 +42,7 @@ If you want to use a string in your code that can't be translated, please use:
     <!-- TODO: Move this to a proper string resource file and collect all other strings associated with this feature once it is stable. -->
     <string name="schedule_reminders_do_not_translate" maxLength="28" comment="Do not translate this string, this feature is still in development. Name of the screen that appears when review reminders are being scheduled.">Schedule reminders</string>
+    <string name="manually_grant_permissions_do_not_translate" comment="Do not translate this string, this feature is still in development. Pop-up message that is shown to the user when they want to grant an app permission that must be granted via the OS settings screen.">This permission must be manually granted from Settings</string>
+    <string name="revoke_permissions_do_not_translate" comment="Do not translate this string, this feature is still in development. Pop-up message that is shown to the user when they are revoking an app permission.">This permission must be manually revoked from Settings</string>
```
This doesn't seem to be a dev only feature

ericli3690 replied:
Oop, good point. I've moved them to 10-preferences, let me know if someplace else would be better!

ericli3690 replied:
Found a better place in 01-core. I'll group these, and the rest of the strings I need for this feature, together with the strings used on the first-time-install permissions screen.

Arthur-Milchior reviewed (commented):
My first remark is that this commit is too big.
For example, `setOnSwitchClickListener` is called in three places, so I think that a first commit that add the parameter to the listener would be an acceptable commit by itself.

Something that shows that there is a problem its that your commit message seems like a bunch of unrelated changes. I don't really feel that reading this commit message helped me understand what I was going to review. 

I think your commit message would generally need a first section that explains why you're doing this.

In this case, it would be something such as 

> Android offers each app a way to explains to show to the user which permissions the app request and why. This was not yet used by AnkiDroid. This commit introduce this page. In order to access it, the user must (list the steps. I'm really not sure what need to be done here)

Admitttedly, it'd be better if we could have the image in the commit message, as your PR description makes it very clear how to access this view, but we don't have such a feature, so it'll need to be text, or at worse a link to the image hosted on github.

Then you can explain why, in order to achieve your goal, you have to do other things.

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/permissions/AllPermissionsExplanation.kt:
```diff
@@ -0,0 +1,120 @@
+            ActivityResultContracts.RequestMultiplePermissions(),
+        ) { requestedPermissions ->
+            if (!requestedPermissions.all { it.value }) {
+                // The permission dialog did not show up or the user denied the permission
```
That would make the code more verbose, though. I feel like `requestedPermission.all` is clean enough and that adding special case handling for if the hashmap contains one item or multiple is a bit too complex.

Arthur-Milchior replied:
What would make the code more complex?
Now that I've read more, I see there can indeed be multiple permissions. In which case, I just request that you updated your comment to use "permissions" instead of "permission"

ericli3690 replied:
Thanks for implementing this at https://github.com/ankidroid/Anki-Android/pull/19033!

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/permissions/PermissionItem.kt:
```diff
@@ -109,16 +109,14 @@ class PermissionItem(
-                Timber.i("permission switch pressed")
-                listener.invoke()
-            }
+            Timber.i("Permission switch clicked, triggering listener")
```
Oh true, I'll double check if this method re-setting the listener is actually necessary. I edited it and kept it because it was there initially. I'll also pass `isGranted` into the logger.

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/permissions/PermissionsActivity.kt:
```diff
@@ -51,6 +56,28 @@ class PermissionsActivity : AnkiActivity() {
+                // if the API version is at least 31. We double check the API version to pass the lint check.
+                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
+                    showAllPermissionsExplanation()
+                }
```
Hm, how do I send a report to AnkiDroid? Just via `Timber.e`, or is there some other library call I need to use?

Arthur-Milchior replied:
Oh, this is why you asked about Timber.e
You need to use `CrashReportService.sendExceptionReport`

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/permissions/PermissionsActivity.kt:
```diff
@@ -51,6 +56,28 @@ class PermissionsActivity : AnkiActivity() {
         setContentView(R.layout.permissions_activity)
         setTransparentStatusBar()
 
+        when (intent.action) {
```
I made this PR reuse `PermissionsActivity` because Brayan told me to in order to reduce redundant code.

ericli3690 replied:
Here's the link to Brayan's comment: https://github.com/ankidroid/Anki-Android/pull/18978#pullrequestreview-3066423658

I do find your rationale compelling, though; it's why I initially created a new Activity before Brayan told me to change it. I've brought back the new standalone activity now. It hosts a PermissionsFragment since doing so allows me to reuse some code from the other permission fragments.

ericli3690 commented:
Thanks Arthur for the review. I'll try to break this up into smaller commits. Will also add a more clean immediate explanation of the feature in the description.

Brayan's also let me know that I can delete TiramisuPermissionsFragment, so I'll do that too.

ericli3690 commented:
@Arthur-Milchior I'm going to wait until https://github.com/ankidroid/Anki-Android/pull/19033 is fully merged before I start making edits here, since you rename some of the variables I'm working with etc. and I don't want to make merging too messy. I'll focus on cleaning up my other draft PRs in the meantime.

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/permissions/AllPermissionsExplanation.kt:
```diff
@@ -0,0 +1,120 @@
+class AllPermissionsExplanation : PermissionsFragment(R.layout.all_permissions_explanation) {
+    /**
+     * Attempts to open the dialog for granting a permission. Falls back to opening the OS settings if
+     * the dialog fails to show up. This may happen if the user has previously denied the permission multiple times,
```
Implemented.

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/permissions/AllPermissionsExplanation.kt:
```diff
@@ -0,0 +1,120 @@
+        val recordAudioPermission = view.findViewById<PermissionItem>(R.id.record_audio_permission_item)
+
+        // External storage
+        if (Permissions.canManageExternalStorage(requireContext())) {
```
Thanks for documenting this at https://github.com/ankidroid/Anki-Android/pull/19034!

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/permissions/AllPermissionsExplanation.kt:
```diff
@@ -0,0 +1,120 @@
+        val notificationPermission = view.findViewById<PermissionItem>(R.id.post_notification_permission_item)
+        val recordAudioPermission = view.findViewById<PermissionItem>(R.id.record_audio_permission_item)
+
+        // External storage
```
I've decided to remove these comments. Thanks

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/permissions/AllPermissionsExplanation.kt:
```diff
@@ -0,0 +1,120 @@
+        // Handle headings
+        view
+            .findViewById<View>(R.id.required_heading)
+            .isVisible = externalStoragePermission.isVisible
```
I've now created a small extra variable to make it more readable.

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/permissions/AllPermissionsExplanation.kt:
```diff
@@ -0,0 +1,120 @@
+            .isVisible = externalStoragePermission.isVisible
+        view
+            .findViewById<View>(R.id.optional_heading)
+            .isVisible = (notificationPermission.isVisible || recordAudioPermission.isVisible)
```
Good point, fixed.

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/permissions/PermissionItem.kt:
```diff
@@ -109,16 +109,14 @@ class PermissionItem(
     /**
-     * It should be use to request the permission.
-     * The listener isn't invoked if the permission is already granted
+     * Should be used to request the permission.
```
I see you've implemented this already at https://github.com/ankidroid/Anki-Android/pull/19033, thanks!

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/permissions/PermissionItem.kt:
```diff
@@ -88,17 +88,17 @@ class PermissionItem(
+     * Listener to be called when the switch is clicked.
+     * Provides the current state of the permission as a boolean. Set by [setOnSwitchClickListener].
+     */
+    private var listener: ((Boolean) -> Unit)? = null
```
I see you've renamed it at https://github.com/ankidroid/Anki-Android/pull/19033, thanks!

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/permissions/PermissionsActivity.kt:
```diff
@@ -31,6 +34,8 @@ import com.ichi2.themes.setTransparentStatusBar
 
 /**
  * Screen responsible for getting permissions from the user.
+ * Also the activity that shows up when the user selects a more-info button in the OS permission settings.
```
Good point, fixed.

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/permissions/PermissionsActivity.kt:
```diff
@@ -51,6 +56,28 @@ class PermissionsActivity : AnkiActivity() {
 
+        when (intent.action) {
+            Intent.ACTION_VIEW_PERMISSION_USAGE, Intent.ACTION_VIEW_PERMISSION_USAGE_FOR_PERIOD -> {
+                // These actions can trigger this activity due to an intent filter in the manifest
```
Moved this docstring to the top of AllPermissionsExplanationActivity and implemented your suggested change.

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/permissions/PermissionsUntil29Fragment.kt:
```diff
@@ -36,22 +35,25 @@ class PermissionsUntil29Fragment : PermissionsFragment(R.layout.permissions_unti
+        ) { requestedPermissions ->
+            if (!requestedPermissions.all { it.value }) {
+                // The permission dialog did not show up or the user denied the permission
+                // Offer the ability to manually grant the permission via the OS settings
```
Fixed!

ericli3690 commented:
Reworded the string that explains why we need notification permissions in AnkiDroid. It now reads: `Allows AnkiDroid to send you review reminder notifications if you create them, allows you to view and cancel sync operations`.

ericli3690 commented:
- Reworded the string again. I feel like the sync bit is more important and should go first. It now reads: `Allows you to view and cancel sync operations, allows AnkiDroid to send you review reminder notifications if you create them`.
- Fixed some docstring grammar issues.
- Moved `offerToGrantOrRevokeOnClick` to a protected method in PermissionsFragment rather than keeping it private within AllPermissionsExplanation to aid with NotificationsPermissionFragment, in this PR: https://github.com/ankidroid/Anki-Android/pull/19167

ericli3690 commented:
Non-functional change: Edited a string resource's explanation comment.

david-allison reviewed (changes requested):
The title should not be 'AnkiDroid needs some permissions to work' - we don't need optional permissions

There should be more of a focus on stating that these are optional permissions (if only optional permissions are displayed)

ericli3690 commented:
Thanks for the review!

Hm, regarding your comment about `AnkiDroid needs some permissions to work`: In the upcoming PR https://github.com/ankidroid/Anki-Android/pull/19167 which is based on top of this one, I introduce the string `AnkiDroid works best with these permissions`. What do you think about using that string as the title for the permissions explanation screen?

david-allison commented:
I like it!

david-allison reviewed (approved):
Looks solid, cheers!

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/permissions/AllPermissionsExplanationFragment.kt:
```diff
@@ -0,0 +1,99 @@
+ */
+@RequiresApi(Build.VERSION_CODES.S)
+class AllPermissionsExplanationFragment : PermissionsFragment(R.layout.all_permissions_explanation_fragment) {
+    val externalStoragePermission: PermissionsItem? by lazy { view?.findViewById(R.id.manage_external_storage_permission_item) }
+    val notificationPermission: PermissionsItem? by lazy { view?.findViewById(R.id.post_notification_permission_item) }
+    val recordAudioPermission: PermissionsItem? by lazy { view?.findViewById(R.id.record_audio_permission_item) }
```
I... honestly don't remember why I did this. Thanks for catching this, I've turned them into local variables within `onViewCreated`.

ericli3690 commented:
- Implemented @BrayanDSO's feedback!
- Rebased.
